mitodl/mitxpro
# Course run selection UI, various backend changes

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #168 
Fixes #124 
Part of #122 
Fixes #116 
Fixes #102 

#### What's this PR do?
On the frontend, adds functionality to list courses for a product, and to allow the user to add one of the runs to the basket. On the backend:
 - add some basket validation on checkout
 - add support for storing the selected run and showing it to the user in the basket API
 - clear the basket on a order fulfillment

Enrolling the user is not handled in this PR

#### How should this be manually tested?
Add a program product to a `BasketItem`. Go to `/checkout`. You should see select elements to select runs for each course. You should be able to change the run and select a different one for purchase when you checkout. You will only be able to see which one was selected by looking at the database, at the `CourseRunSelection` model.

Note that there is nothing which adds a row to `CourseRunEnrollment` but there is code to clear the basket including `CourseRunSelection`. So after order fulfillment the runs should effectively disappear.

#### Screenshots

Program
![pr_186_program](https://user-images.githubusercontent.com/863262/57229974-51ad0c00-6fe5-11e9-98b1-a5baa2f578fb.png)


Course run

![pr186_courserun](https://user-images.githubusercontent.com/863262/57229984-5671c000-6fe5-11e9-8a53-6bfadb8f15ee.png)
